consensus/service/src/validators.rs

```diff
@@ -236,279 +186,160 @@ pub mod well_formed_tests {
             &mut rng,
         );
 
-        assert_eq!(is_well_formed(&tx, &ledger), Ok(()));
-    }
-
-    #[test_with_logger]
-    /// `is_well_formed` should reject a transaction that contains an invalid signature.
-    fn is_well_formed_rejects_invalid_signature(_logger: Logger) {
-        let mut rng = Hc128Rng::from_seed([78u8; 32]);
-
-        let sender = AccountKey::random(&mut rng);
-        let recipient = AccountKey::random(&mut rng);
-
-        let mut ledger = create_ledger();
-        let n_blocks = 3;
-        initialize_ledger(&mut ledger, n_blocks, &sender, &mut rng);
-
-        // Choose a TxOut to spend. Only the output of the last block is unspent.
-        let block_contents = ledger.get_block_contents(n_blocks - 1).unwrap();
-        let tx_out = block_contents.outputs[0].clone();
-
-        let mut tx = create_transaction(
-            &mut ledger,
-            &tx_out,
-            &sender,
-            &recipient.default_subaddress(),
-            n_blocks + 1,
-            &mut rng,
-        );
-
-        assert_eq!(is_well_formed(&tx, &ledger), Ok(()));
+        let untrusted = DefaultTxManagerUntrustedInterfaces::new(ledger.clone());
 
-        // Corrupt the signature.
-        tx.signature.ring_signatures[0].key_image = KeyImage::from(77);
+        // This tx_context contains highest_indices that exceed the number of TxOuts in the ledger.
+        let tx_context = TxContext {
+            locally_encrypted_tx: Default::default(),
+            tx_hash: tx.tx_hash(),
+            highest_indices: tx.get_membership_proof_highest_indices(),
+            key_images: tx.key_images(),
+            output_public_keys: tx.output_public_keys(),
+        };
 
-        match is_well_formed(&tx, &ledger) {
-            Err(TransactionValidationError::InvalidTransactionSignature(_e)) => {} // Expected.
-            Err(e) => {
-                panic!(format!("Unexpected error {}", e));
+        match untrusted.well_formed_check(&tx_context) {
+            Ok((current_block_index, _highest_index_proofs)) => {
+                assert_eq!(current_block_index, n_blocks - 1);
+                // TODO: check returned membership proofs.
             }
-            Ok(()) => panic!(),
+            Err(e) => panic!("Unexpected error {}", e),
         }
     }
 
     #[test_with_logger]
-    /// `is_well_formed` should reject a transaction that contains a key image that is in the ledger.
-    fn is_well_formed_rejects_double_spend(_logger: Logger) {
```

Comment:
Checking for spent key images depends on the current ledger state, and so is out-of-scope for the well-formed check.